subprojects/core-api/src/main/java/org/gradle/api/artifacts/repositories/AuthenticationSupported.java

```diff
@@ -122,6 +123,33 @@
      */
     void credentials(Class<? extends Credentials> credentialsType);
 
+    /**
+     * Configures the credentials for this repository that will be provided by the build with a custom identity.
+     * <p>
+     * Credentials will be provided from Gradle properties based on the given identity.
+     * If credentials for this repository can not be resolved and the repository will be used in the current build, then the build will fail to start and point to the missing configuration.
+     * <pre class='autoTested'>
+     * repositories {
+     *     maven {
+     *         url "${url}"
+     *         credentials(PasswordCredentials, 'myRepoIdentity')
+     *     }
+     * }
+     * </pre>
+     * <p>
+     * The following credential types are currently supported for the {@code credentialsType} argument:
+     * <ul>
+     * <li>{@link org.gradle.api.credentials.PasswordCredentials}</li>
+     * <li>{@link org.gradle.api.credentials.AwsCredentials}</li>
+     * </ul>
+     *
+     * @throws IllegalArgumentException if {@code credentialsType} is not of a supported type
+     *
+     * @since 7.1
+     */
+    @Incubating
+    void credentials(Class<? extends Credentials> credentialsType, String identity);
```

Comment:
Should there also be an overload with a provider ` void credentials(Class<? extends Credentials> credentialsType, Provider<String> identity);`?

Reply:
It's a tough question as `ProviderFactory#credentials` accept a `Provider` identity. For completeness, we should include one. On the other side, adding all these new features is a bit convoluted as we could achieve the same result using composition together with the `ProviderFactory` like this:

```
repositories {
    maven {
        //...
        credentials = providers.credentials(PasswordCredentials, 'myIdentity')
    }
}
```

However, there isn't any access to the credentials as a provider. The next best thing would be something like:

```
<T extends Credentials> void setCredentials(Provider<T> credentials);
```

Unfortunately, having a setter with a provider is a sort of anti-pattern with the provider API. So I don't believe it would be wise to use and we can't really turn the getter into a `Property`. 

My take on this issue is adding `credentials(Class, String)` and `credentials(Class, Provider)` doesn't make the code any worst than what available now. It's easy to depreciate if and when access to the credentials via `Property` is available. So I would add the provider variant of the API for completeness and address the other concern as a strategic work given we can benefit from customizing the identity now (and it's a must for organizations with multiple repositories, i.e. 8+ repos).

@big-guy @jjohannes Any thoughts on that?